Feature - filter diagnostics

amanPras opened the pull request:
## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] New user settings is introduced, which removes semantics error from the diagnostics
```
"cobol-lsp.analysis-mode": {
          "type": "string",
          "enum": [
            "BASIC",
            "ADVANCED"
          ],
          "enumDescriptions": [
            "Enables basic syntax checks",
            "Enables advanced analysis including syntactic and semantic checks"
          ],
          "description": "Enables advanced analysis including syntax and semantic checks",
          "default": "ADVANCED"
        }
``` 

Below code would throw no error with BASIC mode
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. testabc.
       ENVIRONMENT DIVISION.
       CONFIGURATION SECTION.
       SPECIAL-NAMES.
       INPUT-OUTPUT SECTION.
       FILE-CONTROL.
       DATA DIVISION.
        FILE SECTION.
        WORKING-STORAGE SECTION.
        copy abc.
        01 REAL pic x .
        LINKAGE SECTION.
       PROCEDURE DIVISION.
        MAIN-PROCESSING SECTION.
         MAINLINE-PARAGRAPH.
             display A-B-C.
             display REAL.
             STOP RUN.
``` 

Below error would throw error at linkage section
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. testabc.
       ENVIRONMENT DIVISION.
       CONFIGURATION SECTION.
       SPECIAL-NAMES.
       INPUT-OUTPUT SECTION.
       FILE-CONTROL.
       DATA DIVISION.
        FILE SECTION.
        WORKING-STORAGE SECTION.
        copy abc.
        01 copy rest.
        LINKAGE SECTION.
       PROCEDURE DIVISION.
           stop run.
``` 
- [x] Dialects extensions are backward compatible

## Checklist:
- [ ] Each of my commits contains one meaningful change
- [ ] I have performed rebase of my branch on top of the development
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/errors/ErrorFinalizerService.java:
```diff
@@ -38,6 +43,137 @@
 public class ErrorFinalizerService {
 
   private final MessageService messageService;
+  private static DiagnosticSensitivity filterDiagnostics = NORMAL;
```
`filterDiagnostics` should be associated with the instance not static.

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/service/delegates/communications/ServerCommunications.java:
```diff
@@ -197,8 +197,8 @@ private void showMessage(MessageType type, String message) {
 
-  private void logMessage(MessageType type, String message) {
-    getClient().logMessage(new MessageParams(type, clean(message)));
+  private void logMessage(MessageType type, String errorpair) {
+    getClient().logMessage(new MessageParams(type, clean(errorpair)));
```
```suggestion
  private void logMessage(MessageType type, String message) {
    getClient().logMessage(new MessageParams(type, clean(message)));
```

amanPras committed:
feat: add diagnostics level

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

# Conflicts:
#	server/dialect-daco/src/main/java/org/eclipse/lsp/cobol/dialects/daco/DaCoMaidProcessor.java

amanPras committed:
chore: remove unnecessary changes

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java:
```diff
@@ -1824,12 +1826,15 @@ private ProcedureName parseProcedureName(CobolParser.ProcedureNameContext proced
 
-  private void throwException(String wrongToken, @NonNull Locality locality, String message) {
+  private void throwException(
+      String wrongToken, @NonNull Locality locality, MessageTemplate messageTemplate) {
```
Wouldn't it be easier to add `%s` to those three messages and use MessageTemplate.of just like anywhere else?

amanPras replied:
Yes, I will adjust these. Thanks!

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/CICSVisitor.java:
```diff
@@ -341,7 +342,7 @@ private void areaBWarning(@NonNull List<Token> tokenList) {
                         throwException(
                             token.getText(),
                             locality,
-                            messageService.getMessage("CobolVisitor.AreaBWarningMsg"))));
+                            MessageTemplate.of("CobolVisitor.AreaBWarningMsg"))));
```
Similarly here `throwException(locality, MessageTemplate.of("CobolVisitor.AreaBWarningMsg", token.getText()))`?

amanPras committed:
Update server/engine/src/main/java/org/eclipse/lsp/cobol/service/delegates/communications/ServerCommunications.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

amanPras committed:
feat: update as per PR comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
feat: add settings to change diagnostics sensitivity

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
test: add test

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
docs: add code doc for better readability

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: make sure the equals and hashcode contract of SyntaxError considers ErrorCode

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: fix wrong usage of assumingThat assertions

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera reviewed (approved):
pending feedback from the user

amanPras committed:
test: rebase with development

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/errors/ErrorFinalizerService.java:
```diff
@@ -121,4 +268,32 @@ private Predicate<SyntaxError> shouldRaise(Set<SyntaxError> processedErrors) {
+   * @param nonFatalErrors
+   * @return true if diagnostics is within the clients diagnostic level, false otherwise
+   */
+  public boolean filterDiagnotics(SyntaxError syntaxError, Set<String> nonFatalErrors) {
```
```suggestion
  public boolean keepDiagnotics(SyntaxError syntaxError, Set<String> toleratedErrors) {
```

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/errors/ErrorFinalizerService.java:
```diff
@@ -35,6 +40,137 @@
 
   private final MessageService messageService;
+  private DiagnosticSensitivity filterDiagnostics = NORMAL;
+  public static final Set<String> NON_FATAL_ERRORS =
```
```suggestion
  public static final Set<String> TOLARATED_ERRORS =
```

amanPras committed:
Update server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/errors/ErrorFinalizerService.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

amanPras committed:
Update server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/errors/ErrorFinalizerService.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

amanPras committed:
fix: adopt as per PR suggestions

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: typo

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
refactor: update settings attribute to 'analysis-mode'

"cobol-lsp.analysis-mode": {
          "type": "string",
          "enum": [
            "BASIC",
            "ADVANCED"
          ],
          "enumDescriptions": [
            "Enables basic syntax checks",
            "Enables advanced analysis including syntactic and semantic checks"
          ],
          "description": "Enables advanced analysis including syntax and semantic checks",
          "default": "ADVANCED"
        }

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

VitGottwald commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -551,6 +551,19 @@
+          ],
+          "enumDescriptions": [
+            "Enables basic syntax checks",
+            "Enables advanced analysis including syntactic and semantic checks"
```
```suggestion
            "Enables advanced analysis including syntax and semantic checks"
```

VitGottwald commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -551,6 +551,19 @@
+            "Enables basic syntax checks",
+            "Enables advanced analysis including syntactic and semantic checks"
+          ],
+          "description": "Enables advanced analysis including syntax and semantic checks",
```
```suggestion
          "description": "Analysis mode",
```

amanPras committed:
Update clients/cobol-lsp-vscode-extension/package.json

Co-authored-by: Vit Gottwald <vit.gottwald@gmail.com>

amanPras committed:
Update clients/cobol-lsp-vscode-extension/package.json

Co-authored-by: Vit Gottwald <vit.gottwald@gmail.com>

amanPras commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -551,6 +551,19 @@
           "description": "List of preprocessors used by the LSPConfig",
           "uniqueItems": true
+        },
+        "cobol-lsp.analysis-mode": {
```
@zeibura Can you please review

amanPras committed:
feat: add telemetry event of analysis-mode on extension init

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: merge development

Signed-off-by: amanprashant <aman.prashant@broadcom.com>
